openwebstart/src/main/java/com/openwebstart/app/icon/ApplicationIconHandler.java

```diff
@@ -0,0 +1,97 @@
+package com.openwebstart.app.icon;
+
+import com.openwebstart.app.Application;
+import com.openwebstart.http.DownloadInputStream;
+import net.adoptopenjdk.icedteaweb.Assert;
+import net.adoptopenjdk.icedteaweb.logging.Logger;
+import net.adoptopenjdk.icedteaweb.logging.LoggerFactory;
+
+import javax.imageio.ImageIO;
+import java.awt.image.BufferedImage;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+public class ApplicationIconHandler {
+
+    private final static Logger LOG = LoggerFactory.getLogger(DownloadInputStream.class);
+
+    private final static ApplicationIconHandler INSTANCE = new ApplicationIconHandler();
+
+    private final ApplicationIconCache applicationIconCache;
+
+    private final BufferedImage defaultIcon32;
+
+    private final BufferedImage defaultIcon64;
+
+    private final BufferedImage defaultIcon128;
+
+    private final BufferedImage defaultIcon256;
+
+    private final BufferedImage defaultIcon512;
+
+    public ApplicationIconHandler() {
+        this.applicationIconCache = new ApplicationIconCache(Executors.newCachedThreadPool());
+        defaultIcon32 = createDefaultIcon(IconDimensions.SIZE_32);
+        defaultIcon64 = createDefaultIcon(IconDimensions.SIZE_64);
+        defaultIcon128 = createDefaultIcon(IconDimensions.SIZE_128);
+        defaultIcon256 = createDefaultIcon(IconDimensions.SIZE_256);
+        defaultIcon512 = createDefaultIcon(IconDimensions.SIZE_512);
+    }
+
+    public BufferedImage getIconOrDefault(final Application application, final IconDimensions dimension) {
+        final Optional<BufferedImage> image = applicationIconCache.get(application, dimension);
+        if (!image.isPresent()) {
+            applicationIconCache.triggerDownload(application, dimension, false);
+        }
+        return image.orElseGet(() -> getDefaultIcon(dimension));
+    }
+
+    public Future<BufferedImage> getIcon(final Application application, final IconDimensions dimension) {
```

Comment:
method is never called

Reply:
Correct. This method should be used when creating shortcuts